Remove Element.scrollByLines() and Element.scrollByPages()

karlcow opened the pull request:
#### 819e12af49a3d44661f118eda128fe30fc5c4a72
<pre>
Remove Element.scrollByLines() and Element.scrollByPages()
<a href="https://bugs.webkit.org/show_bug.cgi?id=264732">https://bugs.webkit.org/show_bug.cgi?id=264732</a>
<a href="https://rdar.apple.com/118647942">rdar://118647942</a>

Reviewed by Simon Fraser and Dan Glastonbury.

Chromium removed this code in 2014. It seems that there is no internal
dependencies on these features.
Thanks to Anne van Kesteren for giving a better understanding of what
needed to be done.

* LayoutTests/fast/css/element-scrollByLines-scrollByPages-obsolete-expected.txt: Removed.
* LayoutTests/fast/css/element-scrollByLines-scrollByPages-obsolete.html: Removed.
* LayoutTests/webgl/2.0.0/resources/webgl_test_files/deqp/temp_externs/gecko_dom.js:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::scrollByUnits): Deleted.
(WebCore::Element::scrollByLines): Deleted.
(WebCore::Element::scrollByPages): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMElementGtk.cpp:
(webkit_dom_element_scroll_by_lines): Made the function no-op.
(webkit_dom_element_scroll_by_pages): Made the function no-op.
* Source/WebKitLegacy/mac/DOM/DOMElement.h: Moved the definitions to the deprecated section.
* Source/WebKitLegacy/mac/DOM/DOMElement.mm: Made the function no-op and moved them to the deprecated section.
(-[DOMElement scrollByLines:]):
(-[DOMElement scrollByPages:]):

Canonical link: <a href="https://commits.webkit.org/272881@main">https://commits.webkit.org/272881@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c647c5dd68a0af80d110744478a0724ccb0da2a3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33370 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12141 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35279 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35996 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30329 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34342 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14488 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9284 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29467 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33844 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29790 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8932 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9062 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29772 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37326 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30301 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30121 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35181 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7124 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33048 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10930 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7741 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9763 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9904 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMElementGtk.cpp:
```diff
@@ -819,22 +819,6 @@ void webkit_dom_element_scroll_into_view_if_needed(WebKitDOMElement* self, gbool
-    g_return_if_fail(WEBKIT_DOM_IS_ELEMENT(self));
-    WebCore::Element* item = WebKit::core(self);
-    item->scrollByPages(pages);
-}
```
Hi, these are public APIs that can never be removed, but they're probably unused so it's probably safe to remove the implementations and just leave the functions empty.

karlcow replied:
@mcatanzaro to be extra-sure, you are proposing the code to be something along this?

```cpp
// Element.scrollByLines and Element.scrollByPages have been removed. On the other hand, 
// these two functions are Public API and shall not be removed. 
// see https://bugs.webkit.org/show_bug.cgi?id=264732
void webkit_dom_element_scroll_by_lines(WebKitDOMElement* self, glong lines) {}
void webkit_dom_element_scroll_by_pages(WebKitDOMElement* self, glong pages) {}
```

mcatanzaro replied:
Right, but you don't need to add comments because the entire file is public API.

I would do:

```
void webkit_dom_element_scroll_by_lines(WebKitDOMElement* self, glong lines)
{
}

void webkit_dom_element_scroll_by_pages(WebKitDOMElement* self, glong pages)
{
}
```

mcatanzaro replied:
Basically at this point it's an established pattern that DOM APIs can get removed/broken at any time. They are all deprecated and have already been removed in the newer API versions, and most likely no applications actually use them, so it's sort of OK for them to break. But we cannot ever remove public symbols from the library, so those stick around even though they won't do anything anymore.

karlcow replied:
done. Thanks.

djg commented on Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMElementGtk.cpp:
```diff
@@ -821,18 +821,10 @@ void webkit_dom_element_scroll_into_view_if_needed(WebKitDOMElement* self, gbool
 
 void webkit_dom_element_scroll_by_lines(WebKitDOMElement* self, glong lines)
```
Should these functions be removed instead of left as NOPs, or will that break the ABI?

karlcow replied:
hehe 
![image](https://github.com/WebKit/WebKit/assets/505230/98f36b15-ca20-47d9-9b7e-f30cc74b68ca)

see the discussion in https://github.com/WebKit/WebKit/pull/20743#discussion_r1399398481

djg reviewed (approved):
:shipit:

annevk commented on Source/WebKitLegacy/mac/DOM/DOMElement.h:
```diff
@@ -89,8 +89,6 @@ WEBKIT_CLASS_DEPRECATED_MAC(10_4, 10_14)
 - (void)blur WEBKIT_AVAILABLE_MAC(10_6);
 - (void)scrollIntoView:(BOOL)alignWithTop WEBKIT_AVAILABLE_MAC(10_5);
 - (void)scrollIntoViewIfNeeded:(BOOL)centerIfNeeded WEBKIT_AVAILABLE_MAC(10_5);
-- (void)scrollByLines:(int)lines WEBKIT_AVAILABLE_MAC(10_5);
-- (void)scrollByPages:(int)pages WEBKIT_AVAILABLE_MAC(10_5);
```
This looks like it's public API given the WEBKIT_AVAILABLE_MAC declaration so we most likely cannot remove this. Given that nobody has touched this file I think that's very likely to be true.

I think we can probably annotate these methods with WEBKIT_CLASS_DEPRECATED_MAC instead and noop them.

karlcow replied:
I think I will close the PR. :)

annevk commented on Source/WebKitLegacy/mac/DOM/DOMElement.mm:
```diff
@@ -400,18 +400,6 @@ - (void)scrollIntoViewIfNeeded:(BOOL)centerIfNeeded
     unwrap(*self).scrollIntoViewIfNeeded(centerIfNeeded);
 }
 
-- (void)scrollByLines:(int)lines
-{
-    WebCore::JSMainThreadNullState state;
-    unwrap(*self).scrollByLines(lines);
-}
-
-- (void)scrollByPages:(int)pages
-{
-    WebCore::JSMainThreadNullState state;
-    unwrap(*self).scrollByPages(pages);
-}
-
```
You have to no-op these as well. These are part of the macOS ABI.

annevk reviewed (commented):
Non-reviewer looks good. Thanks for cleaning this up!

karlcow committed:
Remove Element.scrollByLines() and Element.scrollByPages()
https://bugs.webkit.org/show_bug.cgi?id=264732
rdar://118647942

Reviewed by Simon Fraser and Dan Glastonbury.

Chromium removed this code in 2014. It seems that there is no internal
dependencies on these features.
Thanks to Anne van Kesteren for giving a better understanding of what
needed to be done.

* LayoutTests/fast/css/element-scrollByLines-scrollByPages-obsolete-expected.txt: Removed.
* LayoutTests/fast/css/element-scrollByLines-scrollByPages-obsolete.html: Removed.
* LayoutTests/webgl/2.0.0/resources/webgl_test_files/deqp/temp_externs/gecko_dom.js:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::scrollByUnits): Deleted.
(WebCore::Element::scrollByLines): Deleted.
(WebCore::Element::scrollByPages): Deleted.
* Source/WebCore/dom/Element.h:
* Source/WebKit/WebProcess/InjectedBundle/API/gtk/DOM/WebKitDOMElementGtk.cpp:
(webkit_dom_element_scroll_by_lines): Made the function no-op.
(webkit_dom_element_scroll_by_pages): Made the function no-op.
* Source/WebKitLegacy/mac/DOM/DOMElement.h: Moved the definitions to the deprecated section.
* Source/WebKitLegacy/mac/DOM/DOMElement.mm: Made the function no-op and moved them to the deprecated section.
(-[DOMElement scrollByLines:]):
(-[DOMElement scrollByPages:]):

Canonical link: https://commits.webkit.org/272881@main
